### PR TITLE
Detect file, function and line of a log call

### PR DESCRIPTION
**Pull request by skreutzberger**
Currently just the log level and the message are used for logging based
on the given protocol. Where the logging call is actually happening is
ignored which makes debugging much harder.

To fix that, I changed the protocol and all files that consume it in
the main Vapor project to get the file, function and line of a log call
by using “magic” Swift variables. These values can then be used for a
more sophisticated logging and can be ignored if not used.

The position of these variables is important to get the correct
position of the log call in the source code.

### CODE REVIEW

**Comment by NathanFlurry** [comment]
I think it'd be better to utilize `NSThread.callStackSymbols()` on macOS (since Linux doesn't support it yet) so you don't have to manually put `#file` ,`#line`, etc. every time you log something.

File: Sources/Vapor/Droplet/Droplet.swift [hunk 1]
```diff
@@ -271,7 +271,7 @@ public class Droplet {
         // iterate through any logs that were emitted
         // before the instance of Log was created.
         for item in logs {
-            log.log(item.level, message: item.message)
+            log.log(item.level, message: item.message, path: #file, function: #function, line: #line)
```
Comment (LoganWright) [inline Sources/Vapor/Droplet/Droplet.swift]: These are all going to call `Droplet.swift 274`, etc. Just getting started on the PR, so don't have a solution offhand.

File: Sources/Vapor/Log/Log+Convenience.swift [hunk 2]
```diff
@@ -1,64 +1,94 @@
...
+     
+     - parameter message: String to log
+     - parameter path: String where logging happens, is automatically set on default
+     - parameter function: String where logging happens, is automatically set on default
+     - parameter line: String where logging happens, is automatically set on default
      */
-    public func verbose(_ message: String) {
-        log(.verbose, message: message)
+    //public func verbose(_ message: String) {
+    public func verbose(_ message: String, _
```
Comment (LoganWright) [inline Sources/Vapor/Log/Log+Convenience.swift]: Any reason why function/path have no external argument but line does?

File: Tests/VaporTests/LogTests.swift [hunk 3]
```diff
@@ -13,9 +13,10 @@ private class DummyLogger: Log {
         output = ""
         enabled = LogLevel.all
     }
-
-    func log(_ level: LogLevel, message: String) {
-        output = "\(level.description) \(message)"
+    
+    func log(_ level: LogLevel, message: String, path: String, function: String, line: Int) {
+        output = "level: \(level.description), message: '\(message)', "
+        output += "path: \(path), function: \(function), line: \(line)"
```
Comment (LoganWright) [inline Tests/VaporTests/LogTests.swift]: file here too.

**Comment by LoganWright** [comment]
@NathanFlurry can you elaborate a bit more? Based on above, you should be able to call `drop.warning("foo")` and get the call site info for free since it's got default args.

Should we consider formatting more than inline? I find with logging this much detail inline, the logs are almost impossible to parse. Some options might be newlining the logs, or possibly things like cutting down file path.

**Comment by NathanFlurry** [comment]
@LoganWright I didn't realize using something like `#line` as a default parameter would put the line of where the function was called. I assume the code was preprocessed to be something like `func test(a: Int = #line)` would convert to `func test(a: Int = 1)`. I should do my research next time. 😄

As for formatting logs, I think it'd be nice to be able to define custom formatters with some predefined formats. For example, in production, you'd probably want more detailed logs (including things like dates) compared to development where you just want the file + line number.

File: Sources/Vapor/Log/Log+Convenience.swift (hunk 2 above)
Comment (skreutzberger) [inline Sources/Vapor/Log/Log+Convenience.swift]: Yes, my fault. I am sloppy about that very often. In the end I just want that nothing after `message` is shown in Xcode code completion so maybe just path should get the `_`?

**Comment by skreutzberger** [comment]
@LoganWright about your concern of the increased volume of logging / more stuff to log:

In the end each logger can decide how much it really wants to consume. I just extended the general protocol so that a logger just gets the chance to receive the line, file and function.  Some loggers will ignore it and some will add even more stuff.

For example for the SwiftyBeaver Vapor logger I will add a customizable formatter where you can decide how much you want to log in what format.

File: Tests/VaporTests/LogTests.swift (hunk 3 above)
Comment (skreutzberger) [inline Tests/VaporTests/LogTests.swift]: I will update the pull request and make the use of `_` more consistent ☺️

**Comment by LoganWright** [comment]
@skreutzberger looks good, the only thing I don't see is path/file stuff. I'm wondering if we should use `file` alongside `#file` for consistency instead of path. If you like that, I think it'd be a quick change and we can get this in today 👍

**Comment by skreutzberger** [comment]
Absolutely! I renamed `path` to `file` and will do the same in the SwiftyBeaver core project. It was one of these early hour sins that never got fixed 😀

**Comment by tannernelson** [comment]
Have we determined if the default arguments of `file: String = #file` will resolve to the file in which that method is defined or the file in which it is called?

If it resolves to where it is defined, then we should make the file optional and default to nil. Getting the file where the method is defined is useless or even worse misleading.

**Comment by LoganWright** [comment]
@tannernelson the #file default arg stuff is an old trick. It works at call site, not define site 👍

**Comment by tannernelson** [comment]
Woot. That's awesome. This LGTM then

**Pull request merged**